### PR TITLE
Native gas price estimator

### PR DESCRIPTION
**Pull request by sunce86**
Native gas price estimator based on the:
https://github.com/zsfelfoldi/feehistory/blob/main/docs/feeOracle.md

General structure of the component is very similar to the BlockNative's, except that in the case of Native gas price estimator we don't have the issue with RATE_LIMIT, therefore, multiple components (orderbook, solver) can instantiate this gas estimator at the same time and access the shared resource - client node.

### CODE REVIEW

**Review by bh2smith — COMMENTED** [review]
Holy wow! This is great. I haven't read the entire thing in detail yet, but curious where you got all this from I see the [link](https://github.com/zsfelfoldi/feehistory/blob/main/docs/feeOracle.md) you provided at the top of the file, but if thats the source, then I am assuming you transformed the spec into all this elaborate code. I will try and cross check the details tomorrow.

**Review by bh2smith — APPROVED** [review]
Looks great although it seems a bit strange to rely on "sample" constants. Are these meant to be a good indicator or are they really just "example values"? I suppose this is what you wanted configurable constants for... Did the lazy static suggestion help at all?

**Review by nlordell — APPROVED** [review]
This is quite complex but looks good to me. Just some small comments and nits.

File: src/nativegasestimator.rs [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+    fn default() -> Self {
+        Self {
+            sample_min_percentile: 10.0,
+            sample_max_percentile: 30.0,
+            max_reward_percentile: 20,
+            min_block_percentile: 30.0,
+            max_block_percentile: 60.0,
+            max_time_factor: 128.0,
+            extra_priority_fee_ratio: 0.25,
+            extra_priority_fee_boost: 1559.0,
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs:56]: Was searching for an interesting number.

File: src/nativegasestimator.rs [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+    types::{BlockNumber, U256},
+    Transport,
+};
+
+const CACHED_RESPONSE_VALIDITY: Duration = Duration::from_secs(60);
+
+//rate limit of ethereum L1 nodes
+const RATE_LIMIT: Duration = Duration::from_secs(5);
+
+// Parameters for Native gas price estimator algorithm
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Fixed.

File: src/nativegasestimator.rs [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        let cached_response_clone = cached_response.clone();
+        let params = params.unwrap_or_default();
+
+        //do one calculation to initially populate cache before any request for gas price estimation is received from our users
+        match suggest_fee(transport.clone(), &params).await {
+            Ok(fees) => {
+                // bump cap to be the ~ 2 x base_fee_per_gas (similar as BlockNative does)
+                let fees = fees
+                    .into_iter()
+                    .map(|(time_limit, gas_price)| (time_limit, gas_price.bump_cap(2.0)))
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Makes sense. Fixed.

File: src/nativegasestimator.rs [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+        //spawn task for updating the cached response every RATE_LIMIT seconds
+        let handle = task::spawn(async move {
+            loop {
+                tokio::time::sleep(RATE_LIMIT).await;
+                match suggest_fee(transport.clone(), &params).await {
+                    Ok(fees) => {
+                        // bump cap to be the ~ 2 x base_fee_per_gas (similar as BlockNative does)
+                        let fees = fees
+                            .into_iter()
+                            .map(|(time_limit, gas_price)| (time_limit, gas_price.bump_cap(2.0)))
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Fixed.

File: src/nativegasestimator.rs [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+// meaningful estimates on variable time scales.
+async fn suggest_fee<T: Transport + Send + Sync>(
+    transport: T,
+    params: &Params,
+) -> Result<Vec<(f64, EstimatedGasPrice)>> {
+    let web3 = web3::Web3::new(transport.clone());
+    let fee_history = web3
+        .eth()
+        .fee_history(
+            300.into(),
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Fixed.

File: src/nativegasestimator.rs [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+            None,
+        )
+        .await?;
+
+    // Initialize
+    let mut base_fee = fee_history.base_fee_per_gas.clone();
+    let mut order: Vec<usize> = vec![];
+    for i in 0..fee_history.base_fee_per_gas.len() {
+        order.push(i);
+    }
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Thanks, I was searching for a better way to write this :)

File: src/nativegasestimator.rs [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+
+    // If a block is full then the baseFee of the next block is copied. The reason is that in full blocks the minimal
+    // priority fee might not be enough to get included. The last (pending) block is also assumed to end up being full
+    // in order to give some upwards bias for urgent suggestions.
+    base_fee[fee_history.base_fee_per_gas.len() - 1] *= 9 / 8;
+    for i in (0..=fee_history.gas_used_ratio.len() - 1).rev() {
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Nice. Fixed.

File: src/nativegasestimator.rs [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+        if fee_history.gas_used_ratio[i] > 0.9 {
+            base_fee[i] = base_fee[i + 1];
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs:195]: Added check.

File: src/nativegasestimator.rs [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+        .unwrap();
+
+        //native gas estimator
+        let native_gas_estimator = NativeGasEstimator::new(transport, None).await.unwrap();
+
+        //blocknative gas estimator
+        let mut header = http::header::HeaderMap::new();
+        header.insert(
+            "AUTHORIZATION",
+            http::header::HeaderValue::from_str("093e3f9c-bd8c-43c6-a6cd-3540fb3a2b70").unwrap(), //or replace with api_key
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Replaced.

File: src/nativegasestimator.rs [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+    use std::time::Duration;
+    use std::{fs::File, io::Write};
+
+    #[tokio::test]
+    #[ignore]
+    async fn real_request() {
+        let mut file = File::create("foo.txt").unwrap();
+
+        let transport = web3::transports::Http::new(
+            "https://mainnet.infura.io/v3/3b497b3196e4468288eb5c7f239e86f4",
```
[the remark this answers is not in the document]
Comment (sunce86) [reply src/nativegasestimator.rs]: Replaced.

**Pull request merged**